src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+        p = Procedure(self._rewrite_node(update))
+
+        return p, self._forward_replace(weakref.ref(p))
+
+    def _forward_replace(self, new_proc):
+        assert self._path[-1][1] is None
+
+        def fwd_node(_):
+            raise InvalidCursorError('cannot forward replaced nodes')
+
+        def fwd_gap(_):
+            assert False, "should never get here"
+
+        def fwd_sel(_):
+            assert False, "should never get here"
+
+        return self._make_forward(new_proc, fwd_node, fwd_gap, fwd_sel)
+
+
+@dataclass
+class Gap(Cursor):
+    _path: list[tuple[str, int]]
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (implementation)
+    # ------------------------------------------------------------------------ #
+
+    def parent(self) -> Node:
+        assert self._path
+        return Node(self._proc, self._path[:-1])
+
+    def before(self, dist=1) -> Node:
+        return self._translate(Node, self._path, -dist)
+
+    def after(self, dist=1) -> Node:
+        return self._translate(Node, self._path, dist - 1)
+
+    def prev(self, dist=1) -> Gap:
+        return self._translate(Gap, self._path, -dist)
+
+    def next(self, dist=1) -> Gap:
+        return self._translate(Gap, self._path, dist)
+
+    # ------------------------------------------------------------------------ #
+    # AST mutation
+    # ------------------------------------------------------------------------ #
+
+    def _insert(self, stmts: list, policy=ForwardingPolicy.AnchorPre):
+        """
+        This is an UNSAFE internal function for inserting a list of nodes at a
+        particular gap in an AST block and providing a forwarding function as
+        collateral. It is meant to be package-private, not class-private, so it
+        may be called from other internal classes and modules, but not from
+        end-user code.
+        """
+        assert self._path
+
+        def update(node):
+            attr, i = self._path[-1]
+            children = getattr(node, attr)
+            return node.update(**{attr: children[:i] + stmts + children[i:]})
+
+        from .API import Procedure
+        p = Procedure(self._rewrite_node(update))
+
+        forward = self._forward_insert(weakref.ref(p), len(stmts), policy)
+        return p, forward
+
+    def _forward_insert(self, new_proc, ins_len, policy):
+        ins_idx = self._path[-1][1]
+
+        def fwd_node(i):
+            return i + ins_len * (i >= ins_idx)
+
+        if policy == ForwardingPolicy.AnchorPre:
+            def fwd_gap(i):
+                return i + ins_len * (i > ins_idx)
+        elif policy == ForwardingPolicy.AnchorPost:
+            def fwd_gap(i):
+                return i + ins_len * (i >= ins_idx)
+        else:
+            def fwd_gap(i):
+                if i == ins_idx:
+                    raise InvalidCursorError('insertion gap was invalidated')
+                return i + ins_len * (i > ins_idx)
```

Comment:
I think it would be ok to remove this policy knob and just have the forwarding behavior in the event of an insertion always be to invalidate the cursor.

Reply:
I'm imagining it will be useful to use the same gap to insert multiple things in a loop... forwarding the gap to after the insertion seems useful. :shrug: